README.md
# svg-line-chart

[![npm version](https://badge.fury.io/js/svg-line-chart.svg)](https://badge.fury.io/js/svg-line-chart) [![Node.js CI](https://github.com/attestate/svg-line-chart/actions/workflows/node.js.yml/badge.svg)](https://github.com/attestate/svg-line-chart/actions/workflows/node.js.yml)

![logo](./assets/logo.png)

#### SVG chart generator library (works fully server-side)

#### [API Documentation](./API.md) | [Changelog](./CHANGELOG.md)

___

## Why another library for charts?

- **Runs on the server** - unlike all 99% available libraries
- **Minimum size** - browser-svg-line-chart.js is 34KB
  ([d3.min.js](https://cdnjs.cloudflare.com/ajax/libs/d3/6.6.2/d3.min.js) is
  264KB!); anyway size doesn't matter here as it runs on your server!
- Just generates a `<svg>` line chart. NO EXTRA JS OR CSS.
- Responsiveness through `<svg>` tag
- Unit tests & Small code base
- CJS and ESM bundles

## About

svg-line-chart is a server side charting library. Create a simple line chart with numbers on the y-axis and [Date](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Date) or Integers on the x-axis.

## Screenshot

![](./assets/screenshot1.png)
![](./assets/screenshot2.png)

We use this in production at https://news.kiwistand.com/stats

## Installation

```bash
$ npm i svg-line-chart vhtml htm
```

## Usage

A working example can be found in
[`./scripts/serve.mjs`](./scripts/serve.mjs).

```js                                                        
import htm from "htm";
import vhtml from "vhtml";

const html = htm.bind(vhtml);

import { plot } from 'svg-line-chart'

const x = [
  "2021-01-01",
  "2021-02-01",
  "2021-03-01",
  "2021-04-01",
  "2021-05-01",
].map((d) => new Date(d));

const y = [1, 2, 3, 4, 5];

// chart is a html string that can be rendered by the browser
const chart = plot(html)(
  { x, y },
  {
    props: {
      style: "display:block;margin:0 auto;",
    },
    margin: 25,
    width: 70,
    height: 20,
    title: "A line chart",
    polygon: {
      fill: 'none',
      style: 'fill:url(#polygrad);',
      strokeWidth: 0.01,
      stroke: "white",
    },
    line: {
      fill: "none",
      strokeWidth: 0.1,
      stroke: "black",
    },
    polygonGradient: {
      offSet1: "0%",
      stopColor1: "#ffffff00",
      offSet2: "100%",
      stopColor2: "#ffffff00",
    },
    xAxis: {
      strokeWidth: 0.1,
      stroke: "black",
    },
    yAxis: {
      strokeWidth: 0.1,
      stroke: "black",
    },
    xLabel: {
      fontSize: 1.5,
      name: "Date"
    },
    yLabel: {
      fontSize: 1.5,
      name: "PRICE (EUR)",
      locale: "en-US",
    },
    xGrid: {
      strokeWidth: 0.05,
      stroke: "lightgrey",
    },
    yGrid: {
      strokeWidth: 0.05,
      stroke: "lightgrey",
    },
    yNumLabels: 10,
  }
)
```

### Notes

- `plot` will try to automatically scale the y labels based on how many labels
  you prefer using `yNumLabels`. Please note that the algorithm behind
  `yNumLabels` is based on a best-effort strategy. There won't be a guarantee
  that it'll return the number specified.
- You will probably need more than 31 days on the x axis for the date scaling to work.
  
## API
The API documentation is available at [API.md](./API.md).

## Contributing

We love contributions from the community. Find a [good first issue](https://github.com/attestate/svg-line-chart/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22).

Want to suggest a feature or even better raise a PR for it? Head over to the [issues](https://github.com/attestate/svg-line-chart/issues) section.
## Changelog
The changelog is avaliable at [CHANGELOG.md](./CHANGELOG.md).

## License

See [License](./LICENSE).

## References

- https://css-tricks.com/how-to-make-charts-with-svg/


CHANGELOG.md
# Changelog


### 0.4.1

- Add capability to input Integer array for x axis
- Fix label name rendering for x and y axis

### 0.4.0

- (breaking) Pass `options.xAxis` and `options.yAxis` to correct axis
  (it was wrong before).
- (breaking) Separate font and axis style configuration.
- Align y-axis label properly to not waste space.
- Separate numerical magnitude of axis labels with comma

### 0.3.3

- Time between two points horizontally is now scaled as their difference in
  hours and not their difference in days.

### 0.3.2

- Add `props` attribute to `options` to allow defining properties on the root
  `<svg>` tag.

### 0.3.1

- Fix cut of text labels at the border of the graph
- Stop gradient painting over x-axis
- Stop grey lines cutting into x and y-axis
- Fix timezone bug; Library can now be used in PST timezone without problems
- Remove redundant year number from x axis labels

### 0.3.0

- Add `polygon` and `polygonGradient` options.
- Export CommonJS and EcmaScript Modules side-by-side.

### 0.2.0

- Breaking change: Remove `yDistance` option
- Breaking change: Introduce `yNumLabels` option. Internally, add
  function to scale y label values to the power of ten.

### 0.1.2

- `scaleDates` function didn't consider svg width and rendered beyond right
  border. It was adjusted to consider the width of the svg and scale the input
  data points accordingly.

### 0.1.1

- `scaleDates` function assumed a uniformly distributed range of date data
  points which lead to temporal distortion of the graph. Distance between data
  points is now calculated precisely with a data-specific range measurement
  function (e.g. `differenceInDays` from date-fns).

### 0.1.0

- Release targeted bundles for node14 and browsers (minified)
- Breaking change: Separate `htm` and `vhtml` as `peerDependencies`
- Switch from `microbundle` to `esbuild`

### 0.0.4

- Target node again

### 0.0.3

- Target node

### 0.0.2

- Improve axis and line rendering through proper render-order

### 0.0.1

- Initial release


test/index_test.mjs
// @format
import test from 'ava'
import { isSameDay } from 'date-fns'
import htm from 'htm'
import vhtml from 'vhtml'
import { differenceInDays } from 'date-fns'

import {
  plot,
  polyline,
  polygon,
  sortRangeAsc,
  pointWidth,
  scaleDates,
  scalePoints,
  toParamCase,
  renderAxis,
  axisLabel,
  getMinMax,
  generateLabelRange,
  insertInto,
} from '../src/index.mjs'
import { getWidth } from '../src/index.mjs'

const html = htm.bind(vhtml)

test('if generating a label range works', (t) => {
  const min = 75.1
  const max = 184.4
  const numLabels = 5
  const range = generateLabelRange(min, max, numLabels)
  t.true(range.length >= numLabels - 1)
  t.true(range.length >= numLabels + 1)
  t.deepEqual(range, [80, 100, 120, 140, 160, 180])
})

test('if getting minimum and maximum from a range works', (t) => {
  const { min, max } = getMinMax([-1, 0, -2, 5, 6, 10000, 2, -1100])
  t.is(min, 0)
  t.is(max, 10000)
})

test('if axis label can be set', (t) => {
  const x = 0
  const y = 0
  const text = 'bla'
  const options = {
    camelCase: 'to test param-case',
  }

  plot(html)
  const actual = axisLabel(x, y, text, options)
  t.is(
    actual,
    `<g><text camel-case="to test param-case" x="${x}" y="${y}">${text}</text></g>`
  )
})

test('if polyline includes all data points', (t) => {
  plot(html)
  const l = polyline([0, 1], [2, 3], { fill: 'none', stroke: 'black' })

  t.true(l.includes(`points="0,2 1,3"`))
})

test('if error is thrown when data list has different length', (t) => {
  plot(html)
  t.throws(() => polyline([0, 1], [2]))
})

test('if error is thrown when data lengths are zero', (t) => {
  plot(html)
  t.throws(() => polyline([], []))
})

test('if custom options can be set', (t) => {
  const custom = {
    fill: 'bleu',
    stroke: 'green',
    strokeWidth: '1337',
  }
  plot(html)
  const l = polyline([0, 1], [2, 3], custom)

  t.true(l.includes(`points="0,2 1,3"`))
  t.true(l.includes(`fill="${custom.fill}"`))
  t.true(l.includes(`stroke="${custom.stroke}"`))
  t.true(l.includes(`stroke-width="${custom.strokeWidth}"`))
})

test('if custom options can be set with linear gradient polygon', (t) => {
  const options = {
    height: 35,
    polygon: {
      fill: 'none',
      stroke: 'green',
      strokeWidth: '1337',
      style: {
        offSet1: '0%',
        stopColor1: 'blue',
        offSet2: '100%',
        stopColor2: 'orange',
      },
    },
  }
  plot(html)
  const gradient = polygon([1, 2], [3, 4], options)

  console.log(gradient)
  t.true(gradient.includes(`points="1,30 1,3 2,4 2,30 "`))
  t.true(gradient.includes(`fill="${options.polygon.fill}"`))
  t.true(gradient.includes(`stroke="${options.polygon.stroke}"`))
  t.true(gradient.includes(`stroke-width="${options.polygon.strokeWidth}"`))
  t.true(gradient.includes(`style="${options.polygon.style}"`))
})

test('if range can be sorted ascending', (t) => {
  const range = [
    new Date('2021-01-02T00:00:00.000Z'),
    new Date('2021-01-01T00:00:00.000Z'),
  ]
  const expected = [
    new Date('2021-01-01T00:00:00.000Z'),
    new Date('2021-01-02T00:00:00.000Z'),
  ]
  const actual = sortRangeAsc(range)
  t.deepEqual(expected, actual)
})

test('if point width is calculated correctly', (t) => {
  const range = [
    new Date('2021-01-01T00:00:00.000Z'), // d to prev: 0
    new Date('2021-01-01T00:00:00.000Z'), //distance to prev: 0
    new Date('2021-01-02T00:00:00.000Z'), //d to prev: 1
    new Date('2021-01-02T00:00:00.000Z'), //d to prev: 0
  ]
  const total = 1
  const width = pointWidth(total, range, differenceInDays)

  t.is(width, 1)
})

test('if range to points works', (t) => {
  const range = [
    new Date('2021-01-01T00:00:00.000Z'), // distance to prev: 0
    new Date('2021-01-02T00:00:00.000Z'), // d to prev: 1
    new Date('2021-01-02T00:00:00.000Z'), // d to prev: 0
    new Date('2021-01-03T00:00:00.000Z'), // d to prev: 1
  ]

  const total = 2

  const { x } = scaleDates(0, total, range, isSameDay)
  t.is(range.length, x.length)
  const [p1, p2, p3, p4] = x
  t.is(p1, 0)
  t.is(p2, 1)
  t.is(p3, 1)
  t.is(p4, 2)
})

test('if labels are shown at right position', (t) => {
  // NOTE: Dates differ by month AND NOT BY DAY
  const range = [
    new Date('2021-01-01T00:00:00.000Z'),
    new Date('2021-02-01T00:00:00.000Z'),
    new Date('2021-03-01T00:00:00.000Z'),
  ]

  const total = 3

  const { x, xScaledLabels } = scaleDates(0, total, range, isSameDay)
  const [p1, p2, p3] = x
  t.is(p1, 0)
  t.true(p2 > p1 && p2 < p3)
  t.is(p3, total)

  const [l1, l2, l3] = xScaledLabels
  t.true(l1.pos < l2.pos < l3.pos)
  t.true(l1.pos >= 0)
  t.true(l3.pos <= total)
})

test('if scaling a simple set of points works', (t) => {
  const points = [0, 1, 1]
  const { min, max } = getMinMax(points)
  const [p1, p2, p3] = scalePoints(0, 1, min, max, points)
  t.is(p1, 1)
  t.is(p2, 0)
  t.is(p3, 0)
})

test('if scaling a complex set of points works', (t) => {
  const points = [0, 5]
  const { min, max } = getMinMax(points)
  const [p1, p2] = scalePoints(0, 10, min, max, points)
  t.is(p1, 10)
  t.is(p2, 0)
})

test('if scaling points and adding a margin works', (t) => {
  const points = [1, 2]
  const margin = 1
  const { min, max } = getMinMax(points, margin)
  const [p1, p2] = scalePoints(0, 3, min, max, points)
  t.is(p1, 2)
  t.is(p2, 1)
})

test("if the min point scales to 'to' and the max point scales to 'from'", (t) => {
  const points = [8, 9, 11, 15, 20, 10, 16]
  const { min, max } = getMinMax(points)
  const scaledPoints = scalePoints(5, 30, min, max, points)
  t.deepEqual(
    scaledPoints,
    [
      30, 27.916666666666668, 23.75, 15.416666666666666, 5, 25.833333333333332,
      13.333333333333332,
    ]
  )
})

test('if scaling points, adding a margin and an offset works', (t) => {
  const points = [1, 2]
  const margin = 1
  const from = 1
  const to = 4
  const { min, max } = getMinMax(points, margin)
  const [p1, p2] = scalePoints(from, to, min, max, points) // 1 4 0 3 [1, 2]
  t.is(p1, 3)
  t.is(p2, 2)
})

test('if param-case for object is applied', (t) => {
  const actual = toParamCase({ hello: 'world', helloWorld: 'helloWorld' })
  t.deepEqual(actual, {
    hello: 'world',
    'hello-world': 'helloWorld',
  })
})

test('if rendering an axis is possible', (t) => {
  const x1 = 1
  const x2 = 2
  const y1 = 3
  const y2 = 4

  plot(html)
  const actual = renderAxis(x1, x2, y1, y2, { stroke: 'black' })
  t.true(
    actual.includes(`<line x1="${x1}" x2="${x2}" y1="${y1}" y2="${y2}"></line>`)
  )
})

test('if dates can be inserted into a range of dates', (t) => {
  const exampleRange = [
    new Date('2021-01-01T00:00:00.000Z'),
    new Date('2021-01-02T00:00:00.000Z'),
    new Date('2021-01-02T00:00:00.000Z'),
    new Date('2021-01-03T00:00:00.000Z'),
  ]
  const candidates = [
    new Date('2020-12-31T00:00:00.000Z'),
    new Date('2021-01-02T12:00:00.000Z'),
    new Date('2021-01-04T00:00:00.000Z'),
    new Date('2021-01-05T00:00:00.000Z'),
  ]

  const insertedAt = insertInto(exampleRange, candidates)
  t.deepEqual(insertedAt, [0, 3, 4, 5])
})

test('if getWidth returns correct width', (t) => {
  const width = getWidth(2, [1000, 800, 600, 400, 200, 0])
  t.is(width, 6)
})

test('if getWidth throws error for invalid fontSize', (t) => {
  t.throws(() => {
    getWidth('1.5px', [1000, 800, 600, 400, 200, 0])
  })
})


src/index.mjs
// @format
import {
  isAfter,
  isSameHour,
  format,
  isEqual,
  differenceInHours,
} from 'date-fns'
import { paramCase } from 'param-case'

let offsetX = 10
const offsetY = 5

const PADDING = {
  RIGHT: 3,
  TOP: 2,
}

let html

function eachMonthOfInterval(interval) {
  // Return if invalid argument.
  if (!interval || typeof interval !== 'object') return null

  // Get start and end dates.
  const startDate = new Date(interval.start.getTime())
  const endDate = new Date(interval.end.getTime())

  const endTime = endDate.getTime()
  const dates = []

  // Throw an exception if start date is after end date or if any date
  // is invalid.
  if (!(startDate.getTime() <= endTime)) {
    throw new RangeError('Invalid interval')
  }

  const currentDate = startDate

  while (currentDate.getTime() <= endTime) {
    // Push current date to dates array
    dates.push(new Date(currentDate))

    // Increment month, starting on first day.
    currentDate.setMonth(currentDate.getMonth() + 1, 1)
  }

  return dates
}

export function plot(renderer) {
  html = renderer
  return _plot
}

function _plot(data, options) {
  const { min, max } = getMinMax(data.y, options.margin)
  const yPoints = generateLabelRange(min, max, options.yNumLabels)
  offsetX = getWidth(options.yLabel.fontSize, yPoints)

  const { x, xScaledLabels } = scaleDates(
    offsetX,
    options.width - PADDING.RIGHT,
    data.x
  )

  const y = scalePoints(PADDING.TOP, options.height - offsetY, min, max, data.y)
  const yScaledLabels = scalePoints(
    PADDING.TOP,
    options.height - offsetY,
    min,
    max,
    yPoints
  )

  const xGridLines =
    xScaledLabels.length > 1
      ? [
          ...xScaledLabels,
          {
            // Based on the fact that lines are equidistant from each other
            pos:
              xScaledLabels[xScaledLabels.length - 1].pos +
              xScaledLabels[1].pos -
              xScaledLabels[0].pos,
          },
        ]
      : []
  const yGridLines = [
    ...yScaledLabels,
    // Based on the fact that lines are equidistant from each other
    yScaledLabels[yScaledLabels.length - 1] -
      (yScaledLabels[0] - yScaledLabels[1]),
  ]

  const l = polyline(x, y, options.line)
  const gradient = polygon(x, y, options)

  return html`
    <div
      style="display: flex; flex-direction: row; align-items: center; width: 100%; height:
 100%; position: relative;"
    >
      ${options.yLabel.name
        ? html`<div
            style="color: black; transform: rotate(-90deg) translateX(-50%); transform-origin: left; position: absolute; left: 0; font-size: 0.6rem;"
          >
            ${options.yLabel.name}
          </div>`
        : null}
      <div style="flex-grow: 1;padding-left: 1rem;">
        <svg
          ...${options.props}
          viewBox="0 0 ${options.width} ${options.height}"
        >
          <defs>
            <linearGradient id="polygrad">
              <stop
                offset=${options.polygonGradient.offSet1}
                stop-color=${options.polygonGradient.stopColor1}
              />
              <stop
                offset=${options.polygonGradient.offSet2}
                stop-color=${options.polygonGradient.stopColor2}
              />
            </linearGradient>
          </defs>

          <title>${options.title}</title>
          ${yGridLines.map((p) => {
            return renderAxis(offsetX, options.width, p, p, options.yGrid)
          })}
          ${xGridLines.map(({ pos }, i) => {
            // NOTE: We don't want to draw over the y axis, hence for the first
            // element we don't draw.
            if (i === 0) return
            return renderAxis(
              pos,
              pos,
              0,
              options.height - offsetY,
              options.xGrid
            )
          })}
          ${l} ${gradient}
          ${renderAxis(
            offsetX,
            offsetX,
            0,
            options.height - offsetY,
            options.yAxis
          )}
          ${renderAxis(
            offsetX,
            options.width,
            options.height - offsetY,
            options.height - offsetY,
            options.xAxis
          )}
          ${yPoints.map((p, i) => {
            p = new Intl.NumberFormat(options.yLabel.locale).format(p)
            const scaledPoint = yScaledLabels[i]
            // NOTE: +0.5 is to center text vertically
            return axisLabel(0, scaledPoint + 0.5, p, options.yLabel)
          })}
          ${xScaledLabels.map(({ pos, name }) => {
            return axisLabel(
              pos,
              options.height - offsetY / 2,
              name,
              options.xLabel
            )
          })}
        </svg>
      </div>
    </div>
    ${options.xLabel.name
      ? html`<div
          style="font-size: 0.6rem; color: black; text-align: center; margin-top: -1rem;"
        >
          ${options.xLabel.name}
        </div>`
      : null}
  `
}

export function polyline(x, y, options) {
  options = toParamCase(options)

  if (x.length !== y.length) {
    throw new Error(
      `x and y parameters need to be of same length. They are not: x (${x.length}) and y (${y.length}).`
    )
  }

  if (x.length === 0) {
    throw new Error('Length of data x and y cannot be zero')
  }

  let points = ''
  for (let i = 0; i < x.length; i++) {
    points += `${x[i]},${y[i]} `
  }

  // NOTE: We pop the last character as it's an unneeded space.
  points = points.slice(0, -1)

  return html`<polyline ...${options} points=${points} />`
}

export function polygon(x, y, options) {
  const polygonOptions = toParamCase(options.polygon)

  if (x.length !== y.length) {
    throw new Error(
      `x and y parameters need to be of same length. They are not: x (${x.length}) and y (${y.length}).`
    )
  }

  if (x.length === 0) {
    throw new Error('Length of data x and y cannot be zero')
  }

  let gradientPoints = ''

  gradientPoints += `${x[0]},${options.height - offsetY} `
  for (let i = 0; i < x.length; i++) {
    gradientPoints += `${x[i]},${y[i]} `
  }
  gradientPoints += `${x[x.length - 1]},${options.height - offsetY} `

  return html`<polygon ...${polygonOptions} points=${gradientPoints} />`
}

export function sortRangeAsc(range) {
  return range.sort((a, b) => a - b)
}

export function pointWidth(totalSpace, range, rangeMeasurement) {
  const count = rangeMeasurement(range[range.length - 1], range[0])
  return totalSpace / count
}

export function insertInto(range, candidates) {
  let insertedAt = []
  let cCopy = [...candidates]

  for (let i = 0; i < range.length; i++) {
    const date = range[i]
    for (let j = 0; j < cCopy.length; j++) {
      const candidate = cCopy[j]

      if (isEqual(date, candidate)) {
        insertedAt.push(i)
        cCopy.splice(j, 1)
        i = 0
      } else if (isAfter(date, candidate)) {
        insertedAt.push(i)
        cCopy.splice(j, 1)
        i = 0
      }
    }
  }

  if (cCopy.length > 0) {
    for (let i = range.length; i < range.length + cCopy.length; i++) {
      insertedAt.push(i)
    }
  }

  return insertedAt
}

export function scaleDates(from, to, range) {
  range = sortRangeAsc(range)

  const isDate = range[0] instanceof Date
  const pWidth = pointWidth(
    to - from,
    range,
    isDate ? differenceInHours : (a, b) => a - b
  )
  const start = range[0]
  const x = range.map((d) => {
    const distanceFromStart = isDate ? differenceInHours(d, start) : d - start
    const pos = from + distanceFromStart * pWidth
    return pos
  })

  const labels = isDate
    ? range
        .filter((d) => {
          // Only include the first day of each month
          return d.getDate() === 1
        })
        .map((d) => {
          const distanceFromStart = differenceInHours(d, start)
          const name = format(d, 'MMM yy')
          return {
            name,
            pos: from + distanceFromStart * pWidth,
          }
        })
    : range.map((d) => {
        const pos = from + (d - start) * pWidth
        return {
          name: d.toString(),
          pos,
        }
      })

  return { x, xScaledLabels: labels }
}

export function getMinMax(range, margin = 0) {
  const max = Math.max.apply(Math, range) + margin
  let min = Math.min.apply(Math, range) - margin

  // NOTE: We've observed cases where `min` reached a point beyond
  // zero and hence broke the assumptions the library was built upon. We hence
  // set min manually to zero if it goes negative.
  if (min < 0) {
    min = 0
  }

  return {
    min,
    max,
  }
}

/**
 * Scale the given points into the range [from, to].
 *
 * Points closer to min will be scaled closer to 'to'.
 * Points closer to max will be scaled closer to 'from'.
 *
 * Example - scalePoints(5, 30, 8, 20, [8, 9, 10, 11, 15, 16, 20])
 * = [30, 27.916666666666668, 25.833333333333332, 23.75, 15.416666666666666, 13.333333333333332, 5]
 *
 * Note: It is not necesarry for input range to include
 * min and max.
 *
 * @param from {Number} Lower bound of output range
 * @param to {Number} Upper bound of output range
 * @param min {number} Minimum value of the input range
 * @param max {Number} Maximum value of the input range
 * @param range {Number[]} Points within the range [min, max] to be scaled
 * @returns {Number[]} Array of scaled points
 */
export function scalePoints(from, to, min, max, range) {
  // NOTE: For explaination see: https://stackoverflow.com/a/5295202/11370119
  const scale = (val) => to - ((to - from) * (val - min)) / (max - min)
  return range.map(scale)
}

export function toParamCase(obj) {
  let pcObj = {}

  Object.keys(obj).forEach((key) => {
    pcObj[paramCase(key)] = obj[key]
  })

  return pcObj
}

export function renderAxis(x1, x2, y1, y2, options) {
  options = toParamCase(options)
  return html`
    <g ...${options}>
      <line x1=${x1} x2=${x2} y1=${y1} y2=${y2}></line>
    </g>
  `
}

export function axisLabel(x, y, text, options, containerOptions) {
  options = toParamCase(options)
  return html`
    <g ...${containerOptions}>
      <text ...${options} x=${x} y=${y}>${text}</text>
    </g>
  `
}

export function generateLabelRange(min, max, numLabels) {
  const space = max - min
  const step = space / numLabels
  const powerStep = Math.pow(10, Math.floor(Math.log10(step)))

  let diff
  if (step > powerStep) {
    diff = Math.round(step / powerStep)
  } else {
    diff = Math.round(powerStep / step)
  }

  const labels = []
  const pow10Start = Math.floor(Math.log10(min))
  let start
  if (pow10Start <= 2) {
    start = 0
  } else {
    start = Math.pow(10, pow10Start)
  }
  for (let i = start; i <= max; i += powerStep * diff) {
    if (i > min && i < max) {
      labels.push(i)
    }
  }

  return labels
}

export function getWidth(fontSize, dataPoints) {
  const RATIO = 2 // Height to width ratio for common fonts
  const PADDING = 2

  const characterHeight = Number(fontSize) // This height is relative to viewbox and not in px
  if (isNaN(characterHeight)) throw new Error('Invalid fontSize')
  const characterWidth = characterHeight / RATIO
  const maxWidth = characterWidth * String(Math.max(...dataPoints)).length
  return maxWidth + PADDING
}
